Add a `test_mstl` module checking if `transform` returns desired components

kcentric opened the pull request:
Closes #6083, resolves #5458.

This PR adds a test file for the [MSTL](https://github.com/sktime/sktime/blob/main/sktime/transformations/series/detrend/mstl.py) class. Currently, all it intends to do is test whether [MSTL](https://github.com/sktime/sktime/blob/main/sktime/transformations/series/detrend/mstl.py)'s `transform` returns the correct components when `return_components` is set to True during initialization.

If it would be beneficial, I could probably expand the testing script to test MSTL in other scenarios (like we do for `_detrender` and `_deseasonalizer`). From @fkiraly's last comment in #5458, I thought this would be the most immediately helpful and nice-to-have test.

There's just one method in my code right now so it should be quick to review. This is my first PR here and I'm still familiarizing myself with the codebase, so any feedback would help.

Couple quick questions (mentioned as a TODO in the code, though they're just notes):

1.) I wonder whether I should test `mstl` using our own `load_airline` or create a fresh small numpy dataset within the script itself? 

2.) For the `assert` statements, should we specify "`Seasonal_3` not returned" for eg, or is it enough to say `Seasonal component not returned"?

Lastly, would it be beneficial to add further tests for other MSTL functionalities (which functionalities)?

Thanks everyone! 😊 🚀

geetu040 commented on sktime/transformations/series/detrend/tests/test_mstl.py:
```diff
@@ -0,0 +1,32 @@
+import numpy as np
+import pandas as pd
+from sktime.datasets import load_airline
+from sktime.transformations.series.detrend import MSTL
```
the import goes like this
```suggestion
from sktime.transformations.series.detrend.mstl import MSTL
```

geetu040 commented on sktime/transformations/series/detrend/tests/test_mstl.py:
```diff
@@ -0,0 +1,32 @@
+
+    # Load our default test dataset
+    series = load_airline()
+    series.index = y.index.to_timestamp()
```
what is the purpose of this line, where is `y` coming from? I think it remains undefined.

geetu040 commented on sktime/transformations/series/detrend/tests/test_mstl.py:
```diff
@@ -0,0 +1,32 @@
+    assert 'resid' in transformed.columns, "Residual component missing"
+    # TODO: Specify "Seasonal_3" and "Seasonal_12" missing for our specific data or not?
+    assert 'seasonal_3' in transformed.columns, "Seasonal component missing" 
+    assert 'seasonal_12' in transformed.columns, "Seasonal component missing"
```
You don't need to give a message like `"Seasonal component missing"` here or in above statements when writing in test cases. This message will not be used anywhere as said here (https://github.com/sktime/sktime/pull/6039#discussion_r1511496318). Non blocking and optional

fkiraly replied:
I actually find messages like these helpful.

More than once I've been digging through test failures, trying to understand what exactly is being tested there. This can be very helpful for other developers!

geetu040 replied:
noted!

fkiraly replied:
here's a recent example, where I turned `assert` into `ValueError`:
https://github.com/astrogilda/tsbootstrap/pull/83/files

Things also got tedious because `assert`-s of equalities do not print out the two sides of the equality, leading to additional debugging overhead.

geetu040 reviewed (changes requested):
I don't think if there is a need to create a fresh time series dataset. You can create one when you need a specific case dataset but here `load_airline` dataset is working just fine. It will be good if you can share why you think there is a need for a fresh time series dataset here

kcentric commented:
Thanks for the notes @geetu040, and exactly @fkiraly, I love descriptive statements for test failures too 🙂. I've fixed the typos now, added the copyright and author name, resolved the little TODOs. 

I suppose it's ready for review now 🙌 . Let me know if I should add additional methods to test MSTL in more scenarios, or proceed to a more important/complex issue.

geetu040 commented:
`Great!` the code looks all good but there are some formatting issues.

I believe it is going to fail some pre-commit checks

You can follow this documentation: https://www.sktime.net/en/stable/developer_guide/coding_standards.html#using-pre-commit

Or do the following

1, Install pre-commit
```
pip install pre-commit
```
2, set up pre-commit
```
pre-commit install
```
3. run commit tests manually or they run automatically when you make a commit
```
pre-commit run --all-files
```

kcentric commented:
I'm sure it's only because one of my comments is ~95 characters instead of
88. Shouldn't be a big deal @geetu040, will fix it in the morning.

On Fri, Mar 8, 2024, 11:23 PM Armaghan ***@***.***> wrote:

> Great! the code looks all good but there are some formatting issues.
>
> I believe it is going to fail some pre-commit checks
>
> You can follow this documentation:
> https://www.sktime.net/en/stable/developer_guide/coding_standards.html#using-pre-commit
>
> Or do the following
>
> 1, Install pre-commit
>
> pip install pre-commit
>
> 2, set up pre-commit
>
> pre-commit install
>
>
>    1. run commit tests manually or they run automatically when you make a
>    commit
>
> pre-commit run --all-files
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sktime/sktime/pull/6084#issuecomment-1986735955>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A43V2KUT7F2MGR56F6IS7MLYXKMGLAVCNFSM6AAAAABEMFFLUCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOBWG4ZTKOJVGU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

geetu040 commented:
I am afraid there still are some formatting issues that need to be cleared first, like for-example
- you are importing `numpy` and `pandas` but not using them
- `__all__ = ["test-MSTL"]` is an necessary line. In Python, the `__all__` variable is used to define a list of public objects of a module.

More checks could be failing here that can be overlooked. You should try `pre-commit` to fix them all

kcentric commented:
The extra imports are just in case we want to add some additional tests. That's one of the considerations I mentioned in the PR

geetu040 commented:
> The extra imports are just in case we want to add some additional tests. That's one of the considerations I mentioned in the PR

In that case we'll import the libraries when needed, but for now before making a commit, it should pass the pre-commit checks. I don't think we can ignore the pre-commit checks in any case. (@fkiraly correct me if I am wrong here)

fkiraly reviewed (changes requested):
Thanks!

We should make sure we pass the linting/pre-commit before merging, see here: https://www.sktime.net/en/stable/developer_guide/coding_standards.html

Not blocking, but I think you could make the failure message more descriptive. Typically, I recommend to add:

* source of the faliurem, i.e., `MSTL`
* specific failure condition

Example: test of `MSTL.transform` failued with `return_components=True`, returned `DataFrame` columns are missing `"transformed"` variable.

fkiraly commented:
Re your questions:

> 1.) I wonder whether I should test mstl using our own load_airline or create a fresh small numpy dataset within the script itself?

I agree with @geetu040, I think it's best if we use existing onboard test datasets.

> 2.) For the assert statements, should we specify "Seasonal_3 not returned" for eg, or is it enough to say `Seasonal component not returned"?

See above, I think specificity is better.

> Lastly, would it be beneficial to add further tests for other MSTL functionalities (which functionalities)?

One test that is typical for interfaced estimators is checking the interfaced module against the `sktime` estimator, e.g., that both return the same values on the same series, to make sure the implemented logic is correct.

kcentric commented:
Thanks @fkiraly and @geetu040. @fkiraly I believe the checks should pass now.

geetu040 reviewed (changes requested):
Are you sure you are running the `pre-commit` right because I think it should still fail some checks.

kcentric commented on sktime/transformations/series/detrend/tests/test_mstl.py:
```diff
@@ -0,0 +1,60 @@
+__author__ = ["krishna-t"]
+
+import pytest
+import statsmodels as statsmodels
```
Yeah, I wanted to remove that one too, but I think it got added automatically when I ran the formatting.

kcentric commented on sktime/transformations/series/detrend/tests/test_mstl.py:
```diff
@@ -0,0 +1,60 @@
+from sktime.transformations.series.detrend.mstl import MSTL
+
+
+@pytest.mark.skipif(
+    statsmodels.tsa.seasonal.MSTL is None, reason="MSTL is not installed"
+)
```
Not sure if that's "better"? Mine is a shorter piece of code, so if mine works then it should be fine unless there's a deeper issue. Let's see what @fkiraly says. Thanks for the suggestion.

geetu040 replied:
There actually is a deeper issue.

We want to skip the test case when `statsmodels` is not installed. And we will have to change the implementation because of the following reasons
1. If `statsmodels` is not installed, the current implementation will raise an error instead of skipping the test.
2. All test cases that involve this scenario use `run_test_for_class` to check soft dependency, so we should keep the implementation coherent where possible.

kcentric replied:
Ah, I see. That makes sense, thanks @geetu040. Though I wonder why the current implementation would raise an error rather than skip the test? Given that  the decorator specially is a `skipif` one. 

Does this have to do with the imports?

geetu040 replied:
> I presume this has something to do with the imports?

Yes you are right, import is the reason

https://github.com/sktime/sktime/pull/6084#issuecomment-1988097443
> The reason for this is because you are using `statsmodels.tsa.seasonal.MSTL` whereas the `MSTL` is not accessible like this, you need to import it separately from the module
> 
> ```python
> from statsmodels.tsa.seasonal import MSTL
> ```
> But this information is irrelevant as `MSTL` from `statsmodels` import is not needed when we are using `run_test_for_class` in `pytest.mark.skipif`

geetu040 commented:
1. `pre-commit` checks are still failing that you can try using
```
pre-commit run --files sktime/transformations/series/detrend/tests/test_mstl.py
```
2. Also the test itself is failing that you should try locally using
```
pytest sktime/transformations/series/detrend/tests/test_mstl.py
```
The reason for this is because you are using `statsmodels.tsa.seasonal.MSTL` whereas the `MSTL` is not accessible like this, you need to import it separately from the module
```py
from statsmodels.tsa.seasonal import MSTL
```
But this information is irrelevant as `MSTL` from `statsmodels` import is not needed when we are using `run_test_for_class` in `pytest.mark.skipif`

fkiraly commented:
starting the tests, will review once they've run through

kcentric commented:
Looking at the results, so `black` is the only one which failed, @fkiraly? Strange, I ran it like 3 or 4 times as far as I remember. Can you see the issue?

kcentric commented:
Formatting fixed now.

geetu040 commented:
> Formatting fixed now.

Not just `blake`, it should pass all the `pre-commit` checks, including `pydocstyle`, which will fail in the current implementation

kcentric commented:
> > Formatting fixed now.
> 
> Not just `blake`, it should pass all the `pre-commit` checks, including `pydocstyle`, which will fail in the current implementation

geetu040 commented:
https://github.com/sktime/sktime/pull/6084#issuecomment-1986735955
> ```
> pre-commit run --all-files
> ```

@kcentric can you please run this command and share the output?

kcentric commented:
Actually @geetu040, I've been trying to get that command to run since the beginning, in both my PyCharm terminal and my Mac terminal. I always end up with a `pre-commit.yaml is not a file` message. What I decided to do was look at the individual checks which were failing in @fkiraly's runs and fix them separately without running `pre-commit` itself. Unfortunately, the downside to that is sometimes the changes getting one test to run might cause another to fail. So I indeed kept trying to get `pre-commit` itself working but still no avail. If you could help me fix that by any chance, then we should be able to resolve this issue like a breeze. 

Also, thanks for your patience guys. This is my first time contributing to a project with rigid formatting guidelines - not that we don't have them over at `Argilla` but they're not so specific. 😅 I suppose that comes with `sktime` a much larger and more complex codebase with many moving parts ..

By next time I should be a pro at following `pre-commit`  guidelines 😁

geetu040 commented:
Thanks for keeping up and sure, no worries, we will figure that out. I'll connect with you in the discord channel.

fkiraly commented:
And thanks for helping, @geetu040! Much appreciated.

kcentric commented:
Tests have passed for me locally, commit successful.

<img width="1025" alt="Screenshot 2024-03-14 at 1 28 14 AM" src="https://github.com/sktime/sktime/assets/121068842/9c2696fe-e7c8-4feb-a387-5758287125ac">

fkiraly reviewed (approved):
Failures are known failures from `main` (see #6094), and this implements a nice addition as discussed.

kcentric commented:
Super!! @fkiraly, and thanks @geetu040 and @yarnabrina (on Discord) for all the insight. 

I'll close this PR for now just so we have things clean, but we could come back later and add some more standardized tests, if that sounds good?

fkiraly commented:
> Wondering if I should close the PR for now or keep it open just so we can come back and add a couple more of the standardized tests if necessary?

No need to close - that would prevent us from merging it.

You can always add tests in other PR - before this is merged, you can branch off the PR branch.
